src/client-agent/start_agent.c

```diff
@@ -140,29 +140,36 @@ void start_agent(int is_startup)
     #endif
     int current_server_id = agt->rip_id;
     while (1) {
-        // (max_retries - 1) attempts
+        // Trying to connect to the server
 
-        for (int attempts = 0; attempts < agt->server[current_server_id].max_retries - 1; attempts++) {
-            if (agent_handshake_to_server(current_server_id, is_startup)) {
+        bool should_enroll = false;
+
+        for (int attempts = 0; attempts < agt->server[current_server_id].max_retries; attempts++) {
+            if (agent_handshake_to_server(current_server_id, is_startup, &should_enroll)) {
                 return;
             }
 
-            sleep(agt->server[current_server_id].retry_interval);
-        }
-
-        // Last attempt
+            if (should_enroll) {
+                break;
+            }
 
-        if (agent_handshake_to_server(current_server_id, is_startup)) {
-            return;
+            if (attempts < agt->server[current_server_id].max_retries - 1) {
+                sleep(agt->server[current_server_id].retry_interval);
+            }
         }
 
         // Try to enroll and extra attempt
 
-        if (agt->enrollment_cfg && agt->enrollment_cfg->enabled) {
-            if (try_enroll_to_server(agt->server[current_server_id].rip, agt->server[current_server_id].network_interface) == 0) {
-                if (agent_handshake_to_server(current_server_id, is_startup)) {
+        if (should_enroll && agt->enrollment_cfg && agt->enrollment_cfg->enabled) {
+            int registration_status = try_enroll_to_server(agt->server[current_server_id].rip, agt->server[current_server_id].network_interface);
+            if (registration_status == 0) {
+                if (agent_handshake_to_server(current_server_id, is_startup, &should_enroll)) {
                     return;
                 }
+            } else {
+                if (registration_status == -2) {
+                    merror_exit("Agent enrollment failed. Invalid enrollment information. Exiting.");
```

Comment:
I suggest changing this behavior to a timeout, even if it's a very long time (an hour or a day). This way, we avoid any situations where a misconfiguration by the manager could shut down agents who wouldn't otherwise be accessible.